eclass/zig.eclass

```diff
@@ -0,0 +1,565 @@
+
+	# Since most arguments in array are also cached by ZBS, we
+	# want to reuse array as much as possible, so prevent
+	# modification of it.
+	declare -g -a -r ZBS_ARGS=(
+		# Base arguments from pkg_setup/setup_base_args
+		"${ZBS_ARGS_BASE[@]}"
+
+		# Arguments from ebuild
+		"${my_zbs_args[@]}"
+
+		# Arguments from user
+		"${ZBS_ARGS_EXTRA[@]}"
+	)
+
+	einfo "ZBS: configured with:"
+	einfo "${ZBS_ARGS[@]}"
+}
+
+# @FUNCTION: zig_src_compile
+# @USAGE: [<args>...]
+# @DESCRIPTION:
+# Calls "ezig build" with previously set ZBS_ARGS.
+# Args passed to this function will be passed after ZBS_ARGS.
+zig_src_compile() {
+	pushd "${BUILD_DIR}" > /dev/null || die
+
+	local args=( "${ZBS_ARGS[@]}" "${@}" )
+	einfo "ZBS: compiling with: ${args[@]}"
+	nonfatal ezig build "${args[@]}" || die "ZBS: compilation failed"
+
+	popd > /dev/null || die
+}
+
+# @FUNCTION: zig_src_test
+# @USAGE: [<args>...]
+# @DESCRIPTION:
+# If "test" step exist, calls "ezig build test" with previously set
+# ZBS_ARGS.
+# Args passed to this function will be passed after ZBS_ARGS.
+# Note: currently step detection might give false positives in
+# very rare cases, it will be improved in the future.
+zig_src_test() {
+	pushd "${BUILD_DIR}" > /dev/null || die
+
+	local args=( "${ZBS_ARGS[@]}" "${@}" )
+
+	# UPSTREAM std.testing.tmpDir and a lot of other functions
+	# do not respect --cache-dir or ZIG_LOCAL_CACHE_DIR:
+	# https://github.com/ziglang/zig/issues/19874
+	mkdir ".zig-cache/" || die
+
+	local found_test_step=false
+
+	local -a steps
+	readarray steps < <(nonfatal ezig build --list-steps "${args[@]}" ||
+		die "ZBS: listing steps failed")
+
+	local step
+	for step in "${steps[@]}"; do
+		# UPSTREAM Currently, step name can have any characters in it,
+		# including whitespaces, so splitting names and descriptions
+		# by whitespaces is not enough for some cases.
+		# We probably need something like  "--list-steps names_only".
+		# In practice, almost nobody sets such names.
+		local step_name="$(awk '{print $1}' <<< "${step}")"
+		if [[ "${step_name}" == test ]]; then
+			found_test_step=true
+			break
+		fi
+	done
+
+	if [[ ${found_test_step} == true ]]; then
+		einfo "ZBS: testing with: ${args[@]}"
+		nonfatal ezig build test "${args[@]}" ||
+			die "ZBS: tests failed"
+	else
+		einfo "Test step not found, skipping."
+	fi
+
+	popd > /dev/null || die
+}
+
+# @FUNCTION: zig_src_install
+# @USAGE: [<args>...]
+# @DESCRIPTION:
+# Calls "ezig build" with DESTDIR and previously set ZBS_ARGS.
+# Args passed to this function will be passed after ZBS_ARGS.
+# Also installs documentation via "einstalldocs".
+zig_src_install() {
+	pushd "${BUILD_DIR}" > /dev/null || die
+	local args=( "${ZBS_ARGS[@]}" "${@}" )
+	einfo "ZBS: installing with: ${args[@]}"
+	DESTDIR="${D}" nonfatal ezig build "${args[@]}" ||
+		die "ZBS: installing failed"
+	popd > /dev/null || die
+
+	pushd "${S}" > /dev/null || die
+	einstalldocs
+	popd > /dev/null || die
```

Comment:
Is there a reason cwd wouldn't be `S`?

```suggestion
	einstalldocs
```

Reply:
```bash
DOCS=( "README.md" "ChangeLog" )

src_install() {
	cd "src/" || die
	zig_src_install
}

Traceback (most recent call last):
  File "/usr/lib/portage/python3.12/doins.py", line 609, in <module>
    sys.exit(main(sys.argv[1:]))
             ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 598, in main
    if _doins(
       ^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 441, in _doins
    return install_runner.install_file(source, os.path.dirname(dest))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 375, in install_file
    return self._ins_runner.run(source, dest_dir)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 183, in run
    sstat = os.stat(source)
            ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: b'README.md'
 * ERROR: sys-fs/ncdu-2.7-r1::gentoo failed (install phase):
 *   dodoc failed

```

`cmake.eclass` also does this

Reply:
I dare say it goes against the principle of least surprised. If I'm switching the current directory before starting the phase, I expect the phase to respect that and not switch it back behind my back.

Reply:
Removed pushd/popd